#19: Add pragma to suppress obsolete warnings

fjmorel opened the pull request:
Fixes #19

I don't know if there is a way to suppress `[Obsolete("comment", true)]`...so I left that for later.

I also considered wrapping _just_ the equality checks in a pragma, in case the generated code uses other obsolete code in the future. But that was more complicated and didn't seem worth it.

diegofrata commented:
Sorry for the delay to get back to you. What .NET SDK version are you using? I uncommented the bits that would trigger CS0619 and surprisingly it works for me. I am running 6.0.101.

fjmorel commented:
Ah, that's because adding `[Obsolete]` to the class/record completely suppresses the property obsoletes. I split that up into a separate test, and also added a commented out Obsolete with custom `DiagnosticId`.

I'm on 6.0.203.

fjmorel commented:
btw, I absolutely love this library. I've been able to remove all our IEquatable implementations. It's also made test-writing so much easier for those records that have collection types on them.

diegofrata commented:
I had a play with the idea of having a SuppressObsoleteWarnings property in the Equatable attribute and accidentally pushed onto your PR. I have reverted it -- apologies if you got any confusing notifications.

I think in the end it's not worth having a separate flag and I assume it would be very difficult for users to find out about its existence. So, basically, your PR exactly as is, is good to go and I am merging it!

There's absolutely nothing we can do about the Obsolete(true) attributes and I think using the IgnoreEquality is the right thing to be used in this situation, as it makes explicit to the user that the property is not used in the comparison in any way. 

Lastly, I don't think Obsolete with custom ids are very commonly used in the wild, but an alternative would be to inspect the Obsolete declaration in each property/class, extract the id from the code and issue the appropriate suppression. I honestly don't think it's worth the effort.

I would like to thank you for your submission and I am happy you find this library useful! I have also been using it heavily in my projects for the same reasons -- it gets rid of a lot of boilerplate, boring and error prone code.

fjmorel commented:
Yay, thank you for merging and releasing this!
